src/api/artifact/controller.go

```diff
@@ -54,10 +60,220 @@ type Controller interface {
 }
 
 // NewController creates an instance of the default artifact controller
-func NewController() Controller {
-	// TODO implement
+func NewController(artMgr artifact.Manager, tagMgr tag.Manager) Controller {
+	return &controller{
+		artMgr: artMgr,
+		tagMgr: tagMgr,
+	}
+}
+
+// TODO handle concurrency, the redis lock doesn't cover all cases
+// TODO As a redis lock is applied during the artifact pushing, we do not to handle the concurrent issues
+// for artifacts and tags？？
+
+type controller struct {
+	artMgr artifact.Manager
+	tagMgr tag.Manager
+}
+
+func (c *controller) Ensure(ctx context.Context, repository *models.RepoRecord, digest string, tags ...string) error {
+	artifact, err := c.ensureArtifact(ctx, repository, digest)
+	if err != nil {
+		return err
+	}
+	for _, tag := range tags {
+		if err = c.ensureTag(ctx, artifact, tag); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ensure the artifact exists under the repository, create it if doesn't exist.
+func (c *controller) ensureArtifact(ctx context.Context, repository *models.RepoRecord, digest string) (*artifact.Artifact, error) {
+	query := &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repository.RepositoryID,
+			"digest":        digest,
+		},
+	}
+	_, artifacts, err := c.artMgr.List(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	// the artifact already exists under the repository, return directly
+	if len(artifacts) > 0 {
+		return artifacts[0], nil
+	}
+
+	// the artifact doesn't exist under the repository, create it first
+	artifact := &artifact.Artifact{
+		ProjectID:    repository.ProjectID,
+		RepositoryID: repository.RepositoryID,
+		Digest:       digest,
+		PushTime:     time.Now(),
+	}
+	// abstract the specific information for the artifact
+	c.abstract(ctx, artifact)
+	// create it
+	id, err := c.artMgr.Create(ctx, artifact)
+	if err != nil {
+		return nil, err
+	}
+	artifact.ID = id
+	return artifact, nil
+}
+
+func (c *controller) ensureTag(ctx context.Context, artifact *artifact.Artifact, name string) error {
+	query := &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": artifact.RepositoryID,
+			"name":          name,
+		},
+	}
+	_, tags, err := c.tagMgr.List(ctx, query)
+	if err != nil {
+		return err
+	}
+	// the tag already exists under the repository
+	if len(tags) > 0 {
+		tag := tags[0]
+		// the tag already exists under the repository and is attached to the artifact, return directly
+		if tag.ArtifactID == artifact.ID {
+			return nil
+		}
+		// the tag exists under the repository, but it is attached to other artifact
+		// update it to point to the provided artifact
+		tag.ArtifactID = artifact.ID
+		tag.PushTime = time.Now()
+		return c.tagMgr.Update(ctx, tag, "ArtifactID", "PushTime")
+	}
+	// the tag doesn't exist under the repository, create it
+	_, err = c.tagMgr.Create(ctx, &tm.Tag{
```

Comment:
Again, concurrent case, would it be easier to create it and return the existing one in case of a dup?

Reply:
See comment above

Reply:
Dup error should handled by this method.

Reply:
Updated